#!/bin/bash

i=$1
j=$2

#trimming
cutadapt -q 20 -m 15 -a AGATCGGAAGAGC -A AGATCGGAAGAGC -o ${i%%?????????}"_trimmed.fq.gz" -p ${j%%?????????}"_trimmed.fq.gz" \
$i $j > ${i%%????????????}"trimming.log"

#cutting
cutadapt -m 15 -u -10 -U 10 -o ${i%%?????????}"_cut.fq.gz" -p ${j%%?????????}"_cut.fq.gz" ${i%%?????????}"_trimmed.fq.gz" \
${j%%?????????}"_trimmed.fq.gz" > ${j%%????????????}"cut.log"

#mapping
bismark --multicore 30 --bam --maxins 500 /PATH/TO/mouse_ref -1 ${i%%?????????}"_cut.fq.gz" -2 ${j%%?????????}"_cut.fq.gz"

#deduplication
deduplicate_bismark -p --bam ${i%%?????????}_cut_bismark_bt2_pe.bam


## {sorting and indexing}
    samtools sort ${i%%?????????}_cut_bismark_bt2_pe.deduplicated.bam ${i%%?????????}_cut_bismark_bt2_pe.deduplicated_sorted
    samtools index ${i%%?????????}_cut_bismark_bt2_pe.deduplicated_sorted


# {Methylation counts}
    bismark_methylation_extractor --buffer_size 20G --gzip --cytosine_report --bedGraph \
    --genome_folder /PATH/TO/mouse_ref ${i%%?????????}_cut_bismark_bt2_pe.deduplicated.bam
